tensorflow/compiler/xla/service/algebraic_simplifier.cc

```diff
@@ -1628,6 +1631,267 @@ StatusOr<HloInstruction*> AlgebraicSimplifierVisitor::OptimizeDotOfGather(
+  // We require that the contracting dimension size is 1
+  if (dnums.lhs_contracting_dimensions_size() != 1 ||
+      dnums.rhs_contracting_dimensions_size() != 1) {
+    return nullptr;
+  }
+  HloInstruction* lhs = dot->mutable_operand(0);
+  HloInstruction* rhs = dot->mutable_operand(1);
+  bool const_on_rhs = false;
+  // We require that the dot has one and only one constant operand.
+  if (lhs->opcode() != HloOpcode::kConstant &&
+      rhs->opcode() == HloOpcode::kConstant) {
+    const_on_rhs = true;
+  } else if (lhs->opcode() == HloOpcode::kConstant &&
+             rhs->opcode() != HloOpcode::kConstant) {
+    const_on_rhs = false;
+  } else {
+    return nullptr;
+  }
+
+  VLOG(10) << "Found Dot reorder contracting dims candidate: "
+           << dot->ToString();
+
+  auto is_eligible_reorder_op = [](
+      const HloInstruction* op, const std::vector<int64>& out_contracting_dims,
+      std::vector<int64>& in_contracting_dims) {
+    if (op->opcode() != HloOpcode::kReshape &&
+        op->opcode() != HloOpcode::kTranspose) {
+      return false;
+    }
+    if (op->opcode() == HloOpcode::kReshape) {
+      // We require that reshape op does not change the non-contracting
+      // dimensions. It is possible to relax this restriction if needed
+      const std::vector<int64> in_dims = op->operand(0)->shape().dimensions();
+      const std::vector<int64> out_dims = op->shape().dimensions();
+      int64 out_contracting_left_idx = out_contracting_dims.front();
+      int64 out_contracting_right_idx = out_contracting_dims.back();
+      for (int i = 0; i < out_contracting_left_idx; i++) {
+        if (out_dims[i] != in_dims[i]) {
+          return false;
+        }
+      }
+      for (int i = out_dims.size() - 1, j = in_dims.size() - 1;
+           i > out_contracting_right_idx; i--, j--) {
+        if (out_dims[i] != in_dims[j]) {
+          return false;
+        }
+      }
+      int64 in_contracting_dims_size =
+          in_dims.size() - out_contracting_left_idx -
+          (out_dims.size() - 1 - out_contracting_right_idx);
+      in_contracting_dims.resize(in_contracting_dims_size);
+      std::iota(in_contracting_dims.begin(), in_contracting_dims.end(),
+                out_contracting_left_idx);
+    } else {  // HloOpcode::kTranspose
+      // We require that transpose only touch contracting dimensions
+      std::vector<int64> transpose_dimensions = op->dimensions();
+      for (int64 i = 0; i < transpose_dimensions.size(); i++) {
+        if (transpose_dimensions[i] != i &&
+            (std::find(out_contracting_dims.begin(), out_contracting_dims.end(),
+                       i) == out_contracting_dims.end() ||
+             std::find(out_contracting_dims.begin(), out_contracting_dims.end(),
+                       transpose_dimensions[i]) ==
+                 out_contracting_dims.end())) {
+          return false;
+        }
+      }
+      in_contracting_dims = out_contracting_dims;
+    }
+    return true;
+  };
+
```

Comment:
Thanks a lot for the through review. I will implement the simple pattern and update the PR soon. As you mentioned, the simple pattern does not completely loose the flexibility of the pattern. It can still catch a chain of transposes followed by a reshape. It just changes the chain tracking from a single pass to multiple passes of this optimization.

The idea of fusing transpose + reshape into the dot makes good sense to me. My understand is that it boils down to implement a general dot with multiple contracting dimensions in arbitrary order (should still be "compact" I think). Even if we can do the layout assignment right, at the moment I don't know if it is simple to implement this gemm_thunk with the equivalent performance as to remove the transpose(s). So I'm planning to implement the simple pattern first and revisit this idea.